File: services/business-services/business-menu-option-v3/application/service/menu_option_service_impl.go
package service

import (
	"business-menu-option-v3/domain/exception"
	"business-menu-option-v3/domain/model/request"
	"business-menu-option-v3/domain/model/response"
	"business-menu-option-v3/infrastructure/mapper"
	"business-menu-option-v3/infrastructure/repository"
	"gorm.io/gorm"
)

type menuOptionServiceImpl struct {
	repository repository.MenuOptionRepository
}

func NewMenuOptionService(menuOptionRepository repository.MenuOptionRepository) MenuOptionService {
	return &menuOptionServiceImpl{
		repository: menuOptionRepository,
	}
}

func (thisService *menuOptionServiceImpl) FindAll() ([]response.MenuOptionResponse, error) {
	menuOptionList, err := thisService.repository.FindAll()
	if err != nil {
		return nil, err
	}
	menuOptionResponseList := make([]response.MenuOptionResponse, len(menuOptionList))
	for i, menuOption := range menuOptionList {
		menuOptionResponseList[i] = mapper.EntityToResponse(menuOption)
	}
	return menuOptionResponseList, nil
}

func (thisService *menuOptionServiceImpl) FindById(id uint) (*response.MenuOptionResponse, error) {
	menuOption, err := thisService.repository.FindById(id)
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, exception.ErrorMenuOptionNotFound
		}
		return nil, err
	}
	menuOptionResponse := mapper.EntityToResponse(*menuOption)
	return &menuOptionResponse, nil
}

func (thisService *menuOptionServiceImpl) Save(request request.MenuOptionRequest) (*response.MenuOptionResponse, error) {
	menuOption := mapper.RequestToEntity(request)
	if err := thisService.repository.Save(&menuOption); err != nil {
		return nil, err
	}
	menuOptionResponse := mapper.EntityToResponse(menuOption)
	return &menuOptionResponse, nil
}

func (thisService *menuOptionServiceImpl) Update(request request.MenuOptionRequest, id uint) (*response.MenuOptionResponse, error) {
	menuOption, err := thisService.repository.FindById(id)
	if err != nil {
		return nil, err
	}
	mapper.UpdateRequestToEntity(request, menuOption)

	if err = thisService.repository.Save(menuOption); err != nil {
		return nil, err
	}

	menuOptionResponse := mapper.EntityToResponse(*menuOption)
	return &menuOptionResponse, nil
}

func (thisService *menuOptionServiceImpl) Delete(id uint) error {
	return thisService.repository.Delete(id)
}

File: services/business-services/menu-v2/src/main/java/com/demo/bbq/business/menuoption/application/service/impl/MenuOptionServiceImpl.java
package com.demo.bbq.business.menuoption.application.service.impl;

import com.demo.bbq.business.menuoption.application.service.MenuOptionService;
import com.demo.bbq.business.menuoption.domain.catalog.MenuCategory;
import com.demo.bbq.business.menuoption.domain.model.request.MenuOptionUpdateRequest;
import com.demo.bbq.business.menuoption.domain.model.request.MenuOptionSaveRequest;
import com.demo.bbq.business.menuoption.domain.model.response.MenuOption;
import com.demo.bbq.business.menuoption.infrastructure.repository.handler.MenuOptionRepositoryHandler;
import io.smallrye.mutiny.Uni;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@ApplicationScoped
public class MenuOptionServiceImpl implements MenuOptionService {

  @Inject
  MenuOptionRepositoryHandler menuOptionRepositoryHandler;

  @Override
  public Uni<List<MenuOption>> findByCategory(String categoryCode) {
    return Optional.ofNullable(categoryCode).isEmpty()
        ? menuOptionRepositoryHandler.findAll()
        : this.validateMenuOptionAndFindByCategory(categoryCode);
  }

  private Uni<List<MenuOption>> validateMenuOptionAndFindByCategory(String categoryCode) {
    MenuCategory.validateCategory.accept(categoryCode);
    return menuOptionRepositoryHandler.findAll()
        .map(menuOptions -> menuOptions.stream().filter(menu -> menu.getCategory().equals(categoryCode))
              .collect(Collectors.toList()));
  }

  @Override
  public Uni<MenuOption> findByProductCode(String productCode) {
    return menuOptionRepositoryHandler.findAll()
        .map(menuOptions -> menuOptions.stream().filter(menuOption -> productCode.equals(menuOption.getProductCode())).findFirst().get());
  }

  @Override
  public Uni<Void> save(MenuOptionSaveRequest menuOptionRequest) {
    return menuOptionRepositoryHandler.save(menuOptionRequest);
  }

  @Override
  public Uni<Void> update(MenuOptionUpdateRequest menuOptionRequest, String productCode) {
    return menuOptionRepositoryHandler.update(productCode, menuOptionRequest);
  }

  @Override
  public Uni<Void> deleteByProductCode(String productCode) {
    return menuOptionRepositoryHandler.deleteByProductCode(productCode);
  }

}
